logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/conversation/keyingmaterials/UpdateKeyingMaterialsUseCase.kt

```diff
@@ -24,9 +26,15 @@ class UpdateKeyingMaterialsUseCaseImpl(
     override suspend fun invoke(): UpdateKeyingMaterialsResult =
         mlsConversationRepository
             .getMLSGroupsRequiringKeyingMaterialUpdate(updateKeyingMaterialThresholdProvider.keyingMaterialUpdateThreshold).map { groups ->
+                var updatesSucceeded = true
                 groups.onEach { groupId ->
-                    mlsConversationRepository.updateKeyingMaterial(groupId)
+                    mlsConversationRepository.updateKeyingMaterial(groupId).onFailure {
+                        updatesSucceeded = false
+                    }
                 }
+                if (updatesSucceeded)
+                    Either.Right(Unit)
+                else Either.Left(Unit)
             }.fold(
                 { UpdateKeyingMaterialsResult.Failure(it) },
                 { UpdateKeyingMaterialsResult.Success }
```

Comment:
Can we simplify this by something like this:

```
groups.map { mlsConversationRepository.updateKeyingMaterial(it) }
  .foldToEitherWhileRight(Unit) { value, _ ->
    value
  }.fold(
    { UpdateKeyingMaterialsResult.Failure(it) },
    { UpdateKeyingMaterialsResult.Success }
  )
```

Reply:
Clean advise! But one question, in that case we stop the whole loop, am I right? My idea is updating as much as we can in each call.

Reply:
Unless Either map is lazy will still update all conversations `foldToEitherWhileRight` will stop on the first error when mapping over a list of results.